ui/index.d.ts

```diff
@@ -154,6 +153,29 @@ export interface AzExtServiceClientCredentials {
     signRequest(webResource: any): Promise<any>;
 }
 
+/**
+ * New T2 Azure SDKs use coreAuth.TokenCredential that have `getToken` rather than
+ * `signRequest`.
+ */
+export interface AzExtServiceClientCredentialsT2 {
+    /**
+     * Gets the token provided by this credential.
+     *
+     * This method is called automatically by Azure SDK client libraries. You may call this method
+     * directly, but you must also handle token caching and token refreshing.
+     *
+     * @param scopes - The list of scopes for which the token will have access.
+     * @param options - The options used to configure any requests this
+     *                TokenCredential implementation might make.
+     */
+     getToken(scopes: string | string[], options?: any): Promise<any | null>;
+}
+
+/**
+ * Loose type to allow for the use of T1 and T2 Azure ARM SDKs
+ */
+export type AzExtServiceClientCredentials = AzExtServiceClientCredentialsT1 | AzExtServiceClientCredentialsT2
```

Comment:
For some requests in the Docker extension we call into `signRequest` (e.g. https://github.com/microsoft/vscode-docker/blob/main/src/utils/azureUtils.ts#L63). It would not be hard to change to `getToken`, but one would only do that if they knew they had to. Should we make a class that still implements `signRequest`?

Alternatively, we could mark `AzExtServiceClientCredentialsT1` deprecated or something like that, or remove it entirely, so that we get a build error, forcing the update to `getToken`.

Reply:
As far as I know, the new credentials from the Azure Account are backwards compatible.  @wwlorey ?

The type `AzExtServiceClientCredentials ` could be either T1 or T2, but that's mostly so `createAzureClient` doesn't scream at us when we use it to lazy load clients.  We could also just have credentials and credentials2 on the `SubscriptionContext` so we can use whichever one typing dictates at the time (though my understanding is that they are actually the same credential? Again @wwlorey)

Reply:
https://github.com/microsoft/vscode-azuretools/blob/main/ui/src/tree/AzureAccountTreeItemBase.ts#L135 Yeah, if you follow credentials down to its base classes, it's => `TokenCredentialsBase` which implements `TokenClientCredentials` that extends `ServiceClientCredentials`.  `ServiceClientCredentials` still has `signRequest` and `TokenClientCredentials` has `getToken`

Reply:
Just to make sure my understanding is correct, it's one object implementing two interfaces, assigned to both `ISubscriptionContext.credentials` and `ISubscriptionContext.credentials2` and the caller can choose which to use based on SDK requirements?

Reply:
Yeah, though I need to do a little work in the account extension to make the ADAL token fully forwards compatible.  As of now, if you use the `createAzureClient` helper, it doesn't seem to always work if you are logged into Azure via ADAL tokens. 

However, I believe that this update should be a non-breaking change and only effect our repos where we upgrade the Azure SDK packages.

Reply:
We discussed some of this in triage yesterday but I wanted to respond here for posterity:

> It would not be hard to change to getToken, but one would only do that if they knew they had to. Should we make a class that still implements signRequest?

We can definitely still support `signRequest` and in fact that work has already been done in the account extension for MSAL (T2 SDK): https://github.com/microsoft/vscode-azure-account/pull/388. When we fix https://github.com/microsoft/vscode-azure-account/issues/415 we'll have to keep support for `signRequest` in mind. 

> As far as I know, the new credentials from the Azure Account are backwards compatible.

`credentials2` has types for both T1 and T2.

> We could also just have credentials and credentials2 on the SubscriptionContext so we can use whichever one typing dictates at the time (though my understanding is that they are actually the same credential?)

[`credentials`](https://github.com/microsoft/vscode-azure-account/blob/30056757606d4fa73f38b9b521f14c34d7455b11/src/azure-account.legacy.api.d.ts#L42) and [`credentials2`](https://github.com/microsoft/vscode-azure-account/blob/30056757606d4fa73f38b9b521f14c34d7455b11/src/azure-account.api.d.ts#L39) are actually different types. And `credentials` will be deprecated when we release 0.10.0 of the extension. It'll still be included in the legacy API though.